### PR TITLE
Add API for getting files from task skeleton

### PR DESCRIPTION
**Pull request by robinmessage**

### CODE REVIEW

File: src/main/java/uk/ac/cam/cl/dtg/teaching/pottery/controllers/TasksController.java [hunk 1]
```diff
@@ -156,6 +162,35 @@ public BuilderInfo pollTaskTestingStatus(String taskId) throws TaskNotFoundExcep
     return taskIndex.getTask(taskId).getTestingCopyBuilderInfo();
   }
 
+  @Override
+  public List<String> listFiles(String taskId, Boolean usingTestingVersion, String variant)
+      throws TaskNotFoundException, TaskMissingVariantException, TaskStorageException {
+    Task t = taskIndex.getTask(taskId);
+    LOG.info("List files " + t + ", " + usingTestingVersion + ", " + variant);
```
Comment (acr31) [inline src/main/java/uk/ac/cam/cl/dtg/teaching/pottery/controllers/TasksController.java]: LOG.info("List files {}, {}, {}",t,usingTestingVersion,variant)

File: src/main/java/uk/ac/cam/cl/dtg/teaching/pottery/task/TaskCopy.java [hunk 2]
```diff
@@ -81,6 +90,39 @@ public File getSolutionLocation(String variant) {
...
+  }
+
+  public List<String> listSkeleton(String variant) throws TaskStorageException {
+    File sourceLocation = getSkeletonLocation(variant);
+    if (!sourceLocation.exists()) {
+      return Collections.emptyList();
+    }
+
+    try {
+      List<String> result = new LinkedList<>();
```
Comment (acr31) [inline src/main/java/uk/ac/cam/cl/dtg/teaching/pottery/task/TaskCopy.java]: ArrayList is better

File: src/main/java/uk/ac/cam/cl/dtg/teaching/pottery/controllers/TasksController.java [hunk 3]
```diff
@@ -156,6 +162,35 @@ public BuilderInfo pollTaskTestingStatus(String taskId) throws TaskNotFoundExcep
     return taskIndex.getTask(taskId).getTestingCopyBuilderInfo();
   }
 
+  @Override
+  public List<String> listFiles(String taskId, Boolean usingTestingVersion, String variant)
```
Comment (acr31) [inline src/main/java/uk/ac/cam/cl/dtg/teaching/pottery/controllers/TasksController.java]: rename to listSkeletonFileNames?

File: src/main/java/uk/ac/cam/cl/dtg/teaching/pottery/controllers/TasksController.java [hunk 4]
```diff
@@ -156,6 +162,35 @@ public BuilderInfo pollTaskTestingStatus(String taskId) throws TaskNotFoundExcep
...
+      if (!c.getInfo().getVariants().contains(variant)) {
+        throw new TaskMissingVariantException("Variant " + variant + " is not defined");
+      }
+
+      return c.listSkeleton(variant);
+    }
+  }
+
+  @Override
+  public Response readFile(String taskId, String fileName, Boolean usingTestingVersion,
```
Comment (acr31) [inline src/main/java/uk/ac/cam/cl/dtg/teaching/pottery/controllers/TasksController.java]: rename to readSkeletonFile

File: src/main/java/uk/ac/cam/cl/dtg/teaching/pottery/controllers/TasksController.java [hunk 5]
```diff
@@ -156,6 +162,35 @@ public BuilderInfo pollTaskTestingStatus(String taskId) throws TaskNotFoundExcep
...
+                           String variant)
+      throws TaskNotFoundException, TaskMissingVariantException {
+    Task t = taskIndex.getTask(taskId);
+    try (TaskCopy c = usingTestingVersion ? t.acquireTestingCopy() : t.acquireRegisteredCopy()) {
+      if (!c.getInfo().getVariants().contains(variant)) {
+        throw new TaskMissingVariantException("Variant " + variant + " is not defined");
+      }
+
+      File filePath = new File(c.getSkeletonLocation(variant), fileName);
```
Comment (acr31) [inline src/main/java/uk/ac/cam/cl/dtg/teaching/pottery/controllers/TasksController.java]: don't you need to sanitize filename here to stop people doing things like ../task.json?

File: src/main/java/uk/ac/cam/cl/dtg/teaching/pottery/controllers/TasksController.java [hunk 6]
```diff
@@ -156,6 +162,35 @@ public BuilderInfo pollTaskTestingStatus(String taskId) throws TaskNotFoundExcep
...
+      return Response.ok(filePath, MediaType.APPLICATION_OCTET_STREAM).build();
```
Comment (acr31) [inline src/main/java/uk/ac/cam/cl/dtg/teaching/pottery/controllers/TasksController.java]: This just returns the file name doesn't it - don't you need something like RepoController.readFile?

File: src/main/java/uk/ac/cam/cl/dtg/teaching/pottery/task/TaskCopy.java [hunk 7]
```diff
@@ -81,6 +90,39 @@ public File getSolutionLocation(String variant) {
     return config.getSolutionDir(copyId, variant);
   }
 
+  public File getSkeletonLocation(String variant) {
+    return config.getSkeletonDir(copyId, variant);
+  }
+
+  public List<String> listSkeleton(String variant) throws TaskStorageException {
```
Comment (acr31) [inline src/main/java/uk/ac/cam/cl/dtg/teaching/pottery/task/TaskCopy.java]: listSkeletonFileNames

File: src/main/java/uk/ac/cam/cl/dtg/teaching/pottery/controllers/TasksController.java (hunk 6 above)
Comment (robinmessage) [reply src/main/java/uk/ac/cam/cl/dtg/teaching/pottery/controllers/TasksController.java]: It seems Response actually knows to return the file contents when given a File. I've added a comment to that effect.

File: src/main/java/uk/ac/cam/cl/dtg/teaching/pottery/controllers/TasksController.java (hunk 5 above)
Comment (robinmessage) [reply src/main/java/uk/ac/cam/cl/dtg/teaching/pottery/controllers/TasksController.java]: Oh yeah, we certainly do. Done.

File: src/main/java/uk/ac/cam/cl/dtg/teaching/pottery/task/TaskCopy.java (hunk 7 above)
Comment (robinmessage) [reply src/main/java/uk/ac/cam/cl/dtg/teaching/pottery/task/TaskCopy.java]: listSkeletonFiles to match elsewhere.

File: src/main/java/uk/ac/cam/cl/dtg/teaching/pottery/task/TaskCopy.java (hunk 2 above)
Comment (robinmessage) [reply src/main/java/uk/ac/cam/cl/dtg/teaching/pottery/task/TaskCopy.java]: I agree; this is code I reused. Will change it.

File: src/main/java/uk/ac/cam/cl/dtg/teaching/pottery/controllers/TasksController.java (hunk 6 above)
Comment (acr31) [reply src/main/java/uk/ac/cam/cl/dtg/teaching/pottery/controllers/TasksController.java]: ah. that's clever - thanks for the comment - that will help future-me

**Pull request merged**